// @framer/guischneider.tabs-to-go/code/Quick_Menu.tsx
import * as React from "react"
import { PropertyControls, addPropertyControls, ControlType } from "framer"

const iconStyle = [
    "ios7",
    "win10",
    "metro",
    "androidL",
    "android",
    "material",
    "material-rounded",
    "material-outlined",
    "color",
    "office",
    "ultraviolet",
    "dotty",
    "nolan",
    "dusk",
]
const iconTitles = [
    "iOS",
    "Windows 10",
    "Windows Metro",
    "Material Design Filled",
    "Android Ice Cream",
    "Material",
    "Material Rounded",
    "Material Outlined",
    "Flat Color",
    "Office",
    "Ultraviolet",
    "Dotted",
    "Gradient Line",
    "Dusk",
]

export class TabsToGo extends React.Component<MenuProps> {
    static propertyControls: PropertyControls = {
        labels: {
            type: ControlType.Array,
            propertyControl: {
                type: ControlType.String,
            },
            title: "Labels",
        },
        style: {
            type: ControlType.Enum,
            options: iconStyle,
            optionTitles: iconTitles,
            title: "Icon Style",
        },
        direction: {
            type: ControlType.SegmentedEnum,
            // defaultValue: "column",
            options: ["column", "row"],
            optionTitles: ["column", "row"],
            title: 'Direction',
        },
        iconLayout: {
            type: ControlType.SegmentedEnum,
            // defaultValue: "row",
            options: ["column", "row"],
            optionTitles: ["column", "row"],
        },
        color: {
            type: ControlType.String,
            title: "Icon Color",
        },
        bgcolor: {
            type: ControlType.Color,
            title: "Background",
        },
        txtcolor: {
            type: ControlType.Color,
            title: "Labels",
        },
    }

    static defaultProps = {
        labels: ["Home", "User", "Search", "Settings"],
        direction: "row",
        iconLayout: "column",
        color: "#FFFFFF",
        txtcolor: "#FFFFFF",
        bgcolor: "#191A1A",
        width: 375,
        height: 50,
        style: "win10",
    }

    //Render Menu Layout
    render() {
        const root: React.CSSProperties = {
            background: `${this.props.bgcolor}`,
            flexDirection: `${this.props.direction}`,
            color: `${this.props.txtcolor}`,
            height: "100%",
            display: "flex",
            justifyContent: "space-around",
            fontSize: 10,
            fontFamily: "-apple-system, BlinkMacSystemFont, sans-serif",
            paddingLeft: 0,
            paddingTop: 4,
        }

        // Declare props
        const iconLayout = this.props.iconLayout
        const color = this.props.color
        const style = this.props.style

        return (
            // Creates MenuItem array
            <div style={root}>
                {this.props.labels.map(function(label) {
                    return (
                        <MenuItem
                            name={label}
                            iconLayout={iconLayout}
                            style={style}
                            color={color}
                        />
                    )
                })}
            </div>
        )
    }
}
// Renders MenuItem
class MenuItem extends React.Component<MenuItemProps> {
    render() {
        const menuItem: React.CSSProperties = {
            display: "flex",
            alignItems: "center",
            flexDirection: `${this.props.iconLayout}`,
        }
        const name = this.props.name
        const color = (this.props.color || "").replace("#", "")
        const style = this.props.style

        return (
            <div style={menuItem}>
                <img
                    src={`https://png.icons8.com/${style}/${name}/${color}/256`}
                    style={icon}
                />
                <span style={menuItemLabel}>{name}</span>
            </div>
        )
    }
}

//Default Styles Override
const icon: React.CSSProperties = {
    background: "transparent",
    height: 28,
    width: 28,
    marginBottom: 4,
}

const menuItemLabel: React.CSSProperties = {
    marginLeft: 0,
}

//Define Props type
interface MenuProps {
    labels: Array<string>
    direction: string
    iconLayout: string
    color: string
    bgcolor: string
    txtcolor: string
    style: string
}

interface MenuItemProps {
    name: string
    iconLayout: string
    style: string
    color: string
}
